setup/productionize/scaling/horizontal-scaling.md

```diff
@@ -0,0 +1,77 @@
+---
+layout: single
+title:  "Horizontally Scale Spinnaker Services"
+sidebar:
+  nav: setup
+---
+
+Spinnaker's services are intended to be replicated for performance and
+availability. This page will explain under which circumstances it can help
+to run more than one copy of any single service, and what considerations
+have to be made when doing so.
+
+It is assumed that the reader can scale/provision more replicas of a service
+either by tweaking the [distributed Kubernetes installation's
+component-sizing](/reference/halyard/component-sizing/#replicas), or via
+whatever automation they chose to deploy Spinnaker services on VMs. The
+mechanics of this process will not be covered here.
+
+> Keep in mind, if you are deploying Spinnaker to Kubernetes, you're more
+> likely to see benefits of horizontal scaling if you've already allocated
+> resource [requests & limits](/reference/halyard/component-sizing) for the
+> services being scaled.
+
+## Scaling Clouddriver
+
+Clouddriver is responsible for caching and retrieving infrastructure, and
+submitting operations to your cloud provider. The former can be quite resource
+intensive, and if you see that calls to
+
+* load cloud infrastructure (e.g. load balancers),
+* perform cache updates (e.g. force cache refresh),
+* search for specific resources (e.g. via the /search endpoint)
+
+are slowing down, it can help to add more replicas of the Clouddriver service.
+
+For a more technical explanation: every copy of Clouddriver tries to (on a
+fixed interval) acquire locks to cache as many shards of your infrastructure as
+possible. The way the shards are partitioned is dependant on your cloud
+provider, and can be inferred from log statements. For example, a GCE
+provider's caching agents will write:
+
+```
+GoogleInfrastructureProvider:my-google-account/europe-west1/GoogleRegionalAddressCachingAgent completed in 0.111s
+```
+
+This indicates that the agent is responsible for a single account, region,
+and resource type.
+
+In addition, as more nodes of Clouddriver are added, the number of reads
+forwarded from the gate service are more evenly distributed as well.
+
+## Scaling Orca
+
+Orca is Spinnaker's execution engine, and manages running pipelines by
+forwarding and awaiting for the status of various tasks requested in a
+pipeline. Central to Orca's orchestration is a queue of messages to handle,
+written into Redis and shared among all Orca nodes. This is explained in more
+detail [in this post on monitoring
+Spinnaker](https://blog.spinnaker.io/monitoring-spinnaker-part-1-4847f42a3abd)
+for the curious.
+
+The key metrics to look for when seeing slowdowns in Spinnaker operations such
+as:
+
+* creating applications,
+* running pipelines,
+* submitting ad-hoc operations (clone, resize, rollback)
+
+relate to this shared queue. In particular, `queue.ready.depth` is the count of
```

Comment:
When you say "relate to..." do you mean the queue contains those metrics?

Reply:
Not quite, reworded to "relate to Orca's ability to process this shared queue" -- does that make sense?

Reply:
Well what I'm trying to say is that all of these operations are processed via this queue. If Orca is processing this queue slowly, these operations will be processed slowly.

Reply:
Something like this? (starting right where you say "The key metrics to look for"


"If you see slowdowns in some of Spinnaker's operations, such as...

* creating applications
* running pipelines
* submitting ad hoc operations (clone, resize, rollback)

...it's likely because Orca is having trouble processing messages from this queue.
One key metric to look at is `queue.ready.depth`, which is the count of messages
that can be processed..."

Reply:
I like it, will change